hyperspace/core/src/macros.rs

```diff
@@ -96,3 +96,797 @@ macro_rules! process_finality_event {
+	($(
+        $(#[$($meta:meta)*])*
+		$name:ident($config:path, $client:path),
+	)*) => {
+		#[derive(Debug, Serialize, Deserialize, Clone)]
+		#[serde(tag = "type", rename_all = "snake_case")]
+		pub enum AnyConfig {
+			$(
+				$(#[$($meta)*])*
+				$name($config),
+			)*
+		}
+
+		#[derive(Clone)]
+		pub enum AnyChain {
+			$(
+				$(#[$($meta)*])*
+				$name($client),
+			)*
+		}
+
+		pub enum AnyFinalityEvent {
+			$(
+				$(#[$($meta)*])*
+				$name(<$client as IbcProvider>::FinalityEvent),
+			)*
+		}
+
+		#[derive(Clone)]
+		pub enum AnyAssetId {
+			$(
+				$(#[$($meta)*])*
+				$name(<$client as IbcProvider>::AssetId),
+			)*
+		}
+
+		#[derive(Debug)]
+		pub enum AnyTransactionId {
+			$(
+				$(#[$($meta)*])*
+				$name(<$client as IbcProvider>::TransactionId),
+			)*
+		}
+
+		#[derive(Error, Debug)]
+		pub enum AnyError {
+			$(
+				$(#[$($meta)*])*
+				#[error("{0}")]
+				$name(<$client as IbcProvider>::Error),
+			)*
+			#[error("{0}")]
+			Other(String),
+		}
+
+		#[async_trait]
+		impl IbcProvider for AnyChain {
+			type FinalityEvent = AnyFinalityEvent;
+			type TransactionId = AnyTransactionId;
+			type Error = AnyError;
+			type AssetId = AnyAssetId;
+
+			async fn query_latest_ibc_events<T>(
+				&mut self,
+				finality_event: Self::FinalityEvent,
+				counterparty: &T,
+			) -> Result<Vec<(Any, Vec<IbcEvent>, UpdateType)>, anyhow::Error>
+			where
+				T: Chain,
+			{
+				match self {
+					$(
+						$(#[$($meta)*])*
+						Self::$name(chain) => {
+							let finality_event = downcast!(finality_event => AnyFinalityEvent::$name)
+								.ok_or_else(|| AnyError::Other("Invalid finality event type".to_owned()))?;
+							chain.query_latest_ibc_events(finality_event, counterparty).await
+						}
+					)*
+				}
+			}
+
+			async fn ibc_events(&self) -> Pin<Box<dyn Stream<Item = IbcEvent> + Send + 'static>> {
+				match self {
+					$(
+						$(#[$($meta)*])*
+						Self::$name(chain) => chain.ibc_events().await,
+					)*
+				}
+			}
+
+			async fn query_client_consensus(
+				&self,
+				at: Height,
+				client_id: ClientId,
+				consensus_height: Height,
+			) -> Result<QueryConsensusStateResponse, Self::Error> {
+				match self {
+					$(
+						$(#[$($meta)*])*
+						Self::$name(chain) => chain
+							.query_client_consensus(at, client_id, consensus_height)
+							.await
+							.map_err(AnyError::$name),
+					)*
+				}
+			}
+
+			async fn query_client_state(
+				&self,
+				at: Height,
+				client_id: ClientId,
+			) -> Result<QueryClientStateResponse, Self::Error> {
+				match self {
+					$(
+						$(#[$($meta)*])*
+						Self::$name(chain) => chain
+							.query_client_state(at, client_id)
+							.await
+							.map_err(AnyError::$name),
+					)*
+				}
+			}
+
+			async fn query_connection_end(
+				&self,
+				at: Height,
+				connection_id: ConnectionId,
+			) -> Result<QueryConnectionResponse, Self::Error> {
+				match self {
+					$(
+						$(#[$($meta)*])*
+						Self::$name(chain) => chain
```

Comment:
It almost feel like you could create another macro for all these very similar methods